File: xion-ios/XionDemo/Data/Repository/XionRepository.swift
import Foundation

protocol XionRepositoryProtocol {
    var sessionManager: SessionManager { get }

    func connect() async throws -> String
    func restoreSession() async -> Bool
    func getBalance() async throws -> BalanceInfo
    func getSbcBalance() async throws -> BalanceInfo
    func getBlockHeight() async throws -> Int64
    func send(toAddress: String, amount: String, memo: String, denom: String) async throws -> TransactionResult
    func executeContract(contractAddress: String, msg: String, funds: String?) async throws -> TransactionResult
    func getVaultBalance() async throws -> BalanceInfo
    func vaultDeposit(amount: String, denom: String) async throws -> TransactionResult
    func vaultWithdraw(amount: String, denom: String) async throws -> TransactionResult
    func vaultWithdrawAll() async throws -> TransactionResult
    func getTx(txHash: String) async throws -> TransactionResult
    func getRecentTransactions(address: String) async throws -> [TransactionResult]
    @MainActor func appendTransaction(_ tx: TransactionResult)
    @MainActor func disconnect()
}

extension XionRepositoryProtocol {
    func send(toAddress: String, amount: String, memo: String) async throws -> TransactionResult {
        try await send(toAddress: toAddress, amount: amount, memo: memo, denom: Constants.coinDenom)
    }
    func vaultDeposit(amount: String) async throws -> TransactionResult {
        try await vaultDeposit(amount: amount, denom: Constants.coinDenom)
    }
    func vaultWithdraw(amount: String) async throws -> TransactionResult {
        try await vaultWithdraw(amount: amount, denom: Constants.coinDenom)
    }
}

File: xion-android-demo/app/src/main/java/com/burnt/xiondemo/data/repository/XionRepository.kt
package com.burnt.xiondemo.data.repository

import com.burnt.xiondemo.data.model.BalanceInfo
import com.burnt.xiondemo.data.model.TransactionResult
import com.burnt.xiondemo.data.model.WalletState
import com.burnt.xiondemo.util.Constants
import com.burnt.xiondemo.util.Result
import kotlinx.coroutines.flow.StateFlow

interface XionRepository {
    val walletState: StateFlow<WalletState>
    val transactionHistory: StateFlow<List<TransactionResult>>

    suspend fun prepareSessionKey(): Result<String>
    suspend fun completeConnection(metaAccountAddress: String): Result<String>
    suspend fun restoreSession(): Result<Boolean>

    suspend fun getBalance(): Result<BalanceInfo>
    suspend fun getSbcBalance(): Result<BalanceInfo>
    suspend fun getBlockHeight(): Result<Long>
    suspend fun send(toAddress: String, amount: String, memo: String, denom: String = Constants.COIN_DENOM): Result<TransactionResult>
    suspend fun executeContract(contractAddress: String, msg: String, funds: String?): Result<TransactionResult>
    suspend fun getTx(txHash: String): Result<TransactionResult>
    suspend fun getRecentTransactions(address: String): Result<List<TransactionResult>>

    fun appendTransaction(tx: TransactionResult)
    fun disconnect()
}
